### ISSUE TITLE
Non-deterministic bootstrap, swingset sensitivity to config.bundles insertion order

### ISSUE THREAD

**Issue by mhofman** [Agoric/Agoric#5226]
## Describe the bug
It seems that the cause of transient failures in the [deployment integration CI job](https://github.com/Agoric/agoric-sdk/actions/workflows/deployment-test.yml) is due to bundles sometimes being added in a different order.

If I had a guess, it'd be that multiple bundles are loaded in parallel during bootstrap, which results in a filesystem race with a different resolution order.

## To Reproduce
No consistent reproduction. I do have a full trace however.

## Expected behavior
Deterministic bootstrap

## Platform Environment
 - what OS are you using? Debian on WSL2

## Screenshots
```diff
--- validator0-swingstore-trace 2022-04-25 19:22:38.390978962 -0700
+++ chain-stage-0-swingstore-trace      2022-04-25 19:22:30.351551207 -0700
@@ -34,6 +34,4 @@
 set bundle.b1-fbbb4cca1bafdacb2d62b8c758687e26e6a26e7b2a8e52668bea7950e1914d5db53e38873dc70b15f7a94fa3258c843164400ec244ce7694add59372389648ef {"moduleFormat":"endoZipBase64","endoZipBase64":"UEsDBAo
 set bundle.b1-4d1ecba749269094203571b57b2c29623c9e3538d8ff1f81619936f949933f52bcd1b8e6a5048a521807c36e7290f1bd61ce689573c505b9306efc903202e779 {"moduleFormat":"endoZipBase64","endoZipBase64":"UEsDBAo
-set bundle.b1-b7c184daa30a846c4d37828a8538692e5f9f516d4955276ff09323d2b213229e611266e5acd35753ab6f87c0baa076546ce3aaf97be8d738c7746a3cf7afeccc {"moduleFormat":"endoZipBase64","endoZipBase64":"UEsDBAo
-set bundle.b1-0e514260bc8b97f9285128d043c25f7efe2d599495bbc885e0bf10ca55c0b4604a3b6cae3323d044b1c426cde2bfd0f2bdcb0a379f1d924ae67e5eb3b3a35c90 {"moduleFormat":"endoZipBase64","endoZipBase64":"UEsDBAo
 set bundle.b1-e8ce1f17cce9febedada5854f7658b60fdf036eb8802a697666de65d200fb202a2470340e6353801c45ae63117ecff7e1168e57d24fc5732a69224e365dbf0c1 {"moduleFormat":"endoZipBase64","endoZipBase64":"UEsDBAo
 set bundle.b1-cb9ac4755d6460dc23b4359b0681c0e9e842e53be0a13b80e6531350b2497634185c56fa0a3294fac4360d8b6ffc68c82973d0036e231c68d4c2bb00c8e82e9d {"moduleFormat":"endoZipBase64","endoZipBase64":"UEsDBAo
@@ -45,5 +43,7 @@
 set bundle.b1-0ff8cd627c9a3890748e662092417c3c76caf947882ae0011a17a875cbebee7c99954d5de6971dc8389827c206e5b493af37dbc8afc6fe661debc5bda6bcf835 {"moduleFormat":"endoZipBase64","endoZipBase64":"UEsDBAo
 set bundle.b1-e939c28a5a76dcdf4e426a7e7c0ef8e1e8d980e9a4ee463abeb113ed1e7de12c20f972ff4815007142f155bfb45366140d89e39bcb47199957869e98b90b575a {"moduleFormat":"endoZipBase64","endoZipBase64":"UEsDBAo
+set bundle.b1-b7c184daa30a846c4d37828a8538692e5f9f516d4955276ff09323d2b213229e611266e5acd35753ab6f87c0baa076546ce3aaf97be8d738c7746a3cf7afeccc {"moduleFormat":"endoZipBase64","endoZipBase64":"UEsDBAo
 set bundle.b1-a356f37afb33bd4092e249b2855c3e2d4d0c8012f6af34a6f3a7d6415c5035e9f48ba2a7febee10cd9ccefa5fb5b1c1f3dc788cae39aa42eed09a5b92760656c {"moduleFormat":"endoZipBase64","endoZipBase64":"UEsDBAo
+set bundle.b1-0e514260bc8b97f9285128d043c25f7efe2d599495bbc885e0bf10ca55c0b4604a3b6cae3323d044b1c426cde2bfd0f2bdcb0a379f1d924ae67e5eb3b3a35c90 {"moduleFormat":"endoZipBase64","endoZipBase64":"UEsDBAo
 set bundle.b1-3171e8704105789fdfbbc389fc84eb6140a9949fdb3a3c533e5b05b5f2225bb50603648ebe5b2118b0b2b6f2bb7838eec2d082157657dc92200adc5cd18d1c6a {"moduleFormat":"endoZipBase64","endoZipBase64":"UEsDBAo
 set bundle.b1-3eea6d8392981bdf4ddf24a0567600bc771502b8798b35bb83d4cb503796ebe783abb664f6cc183ceee8052b3962815e071f72e780b67c8c6f73123878a19e74 {"moduleFormat":"endoZipBase64","endoZipBase64":"UEsDBAo
@@ -69,6 +69,4 @@
 set namedBundleID.walletManager b1-fbbb4cca1bafdacb2d62b8c758687e26e6a26e7b2a8e52668bea7950e1914d5db53e38873dc70b15f7a94fa3258c843164400ec244ce7694add59372389648ef
 set namedBundleID.zoe b1-4d1ecba749269094203571b57b2c29623c9e3538d8ff1f81619936f949933f52bcd1b8e6a5048a521807c36e7290f1bd61ce689573c505b9306efc903202e779
-set namedBundleID.coreProposal1_1 b1-b7c184daa30a846c4d37828a8538692e5f9f516d4955276ff09323d2b213229e611266e5acd35753ab6f87c0baa076546ce3aaf97be8d738c7746a3cf7afeccc
-set namedBundleID.coreProposal1_behaviors b1-0e514260bc8b97f9285128d043c25f7efe2d599495bbc885e0bf10ca55c0b4604a3b6cae3323d044b1c426cde2bfd0f2bdcb0a379f1d924ae67e5eb3b3a35c90
 set namedBundleID.coreProposal0_1 b1-e8ce1f17cce9febedada5854f7658b60fdf036eb8802a697666de65d200fb202a2470340e6353801c45ae63117ecff7e1168e57d24fc5732a69224e365dbf0c1
 set namedBundleID.coreProposal0_2 b1-cb9ac4755d6460dc23b4359b0681c0e9e842e53be0a13b80e6531350b2497634185c56fa0a3294fac4360d8b6ffc68c82973d0036e231c68d4c2bb00c8e82e9d
@@ -80,5 +78,7 @@
 set namedBundleID.coreProposal0_8 b1-0ff8cd627c9a3890748e662092417c3c76caf947882ae0011a17a875cbebee7c99954d5de6971dc8389827c206e5b493af37dbc8afc6fe661debc5bda6bcf835
 set namedBundleID.coreProposal0_9 b1-e939c28a5a76dcdf4e426a7e7c0ef8e1e8d980e9a4ee463abeb113ed1e7de12c20f972ff4815007142f155bfb45366140d89e39bcb47199957869e98b90b575a
+set namedBundleID.coreProposal1_1 b1-b7c184daa30a846c4d37828a8538692e5f9f516d4955276ff09323d2b213229e611266e5acd35753ab6f87c0baa076546ce3aaf97be8d738c7746a3cf7afeccc
 set namedBundleID.coreProposal0_behaviors b1-a356f37afb33bd4092e249b2855c3e2d4d0c8012f6af34a6f3a7d6415c5035e9f48ba2a7febee10cd9ccefa5fb5b1c1f3dc788cae39aa42eed09a5b92760656c
+set namedBundleID.coreProposal1_behaviors b1-0e514260bc8b97f9285128d043c25f7efe2d599495bbc885e0bf10ca55c0b4604a3b6cae3323d044b1c426cde2bfd0f2bdcb0a379f1d924ae67e5eb3b3a35c90
 set vat.name.bootstrap v1
 set v1.o.nextID 50
@@ -197,5 +197,5 @@
 set initialized true
 set local.kernelStats {"syscalls":1,"syscallSend":1,"syscallCallNow":0,"syscallSubscribe":0,"syscallResolve":0,"syscallExit":0,"syscallVatstoreGet":0,"syscallVatstoreSet":0,"syscallVatstoreGetAfter":
-set activityhash dd621044390b4c67901ea18623c462226bcff33bda834acb77533024b393b978
+set activityhash f90ad3a726f5f67b3b641c7c1b766eb37006fb7c457eaa4caf5266aca99ade55
```
[chain-stage-0-swingstore-trace-head.gz](https://github.com/Agoric/agoric-sdk/files/8559432/chain-stage-0-swingstore-trace-head.gz)
[validator0-swingstore-trace-head.gz](https://github.com/Agoric/agoric-sdk/files/8559434/validator0-swingstore-trace-head.gz)

**Comment by warner** [comment]
I suspect that the host application is passing `config.vats` or `config.bundles` with a latent insertion-order that varies from one node to another, because of some race in the way they're created.

I looked through `src/controller/initializeSwingset.js` to understand places where we depend upon variable-duration activity (mostly calls to `bundleSource`) and then iterate over object properties, such that the order in which the activity completes can affect the iteration order.

There are a number of places where swingset is sensitive to the iteration order of what it gets from the host, but I think we're using `bundleSource` in a way that does not cause additional reordering.

I'd like to declare that the kernel depends upon the visible contents of `config` but not the latent insertion order. That means we need to process things like `config.vats`, `config.devices`, and `config.bundles` in sorted-keys order.

I'd also like to put an explicit sort after any use of `Promise.all`, to make it easier to audit that we're insensitive to such things.

So in `src/controller/initializeSwingset.js`, every place where we use `Object.keys` should be examined to determine whether the object being listed is the output of a `Promise.all`. If so, we should sort the results before merging them into the next step.

 At a quick glance, this includes at least:
 * near the start of `initializeSwingset`, when we provide empty defaults for each of `config.vats/bundles/devices`, we should sort the tables at that point
 * `normalizeConfigDescriptor`: where we walk `Object.keys(desc)` to build up a list of normalization jobs, we don't need to sort first, because we're going to sort after
   * `normalizeConfigDescriptor` is called from the start of `loadSwingsetConfigFile`, and modifies `config.vats`/etc in-place
* `processGroup` does `const names = Array.from(Object.keys(group))`, uses that to build a list of `processDesc()` promises, then uses `bundlesWithID = await Promise.all()` to await the whole batch. We do not iterate the `bundlesWithID` that comes out (we iterate `names` instead), so as long as we sorted `config.vats/etc` earlier, `names` should be consistent
* After calling `processGroup` for each group, we merge their returned `idToBundle` records into `kconfig.idToBundle`. By sorting the groups earlier, this should be stable, however it would be nice to have the insertion order be sorted by `bundleID` so that any overlap between the groups doesn't cause confusion.
* After that, we iterate `nameToBundle` to populate `kconfig.namedBundleIDs`. `nameToBundle`'s insertion order will be equal to the insertion order of `config.bundles`, followed by `vats` and `devices`. Again, it would be nice if this were sorted by name.

**Comment by warner** [comment]
Yeah, `extractCoreProposalBundles()` in  https://github.com/Agoric/agoric-sdk/blob/39962808ce77432e6ce74a1bdac207f10545a626/packages/cosmic-swingset/src/extract-proposal.js#L116 is doing a big parallel bunch of `bundleSource` calls, then doing `bundleHandles.add()` in variable order as each one finishes. Then it iterates through `bundleHandle.values()` and populates `bundles[bundleID] =` in that same order. So the `config.bundles` that swingset receives will vary depending upon the timing of the `extractCoreProposalBundles()` finishes its bundling, and the kernel DB writes will vary as a result.

We should fix this in both places: `extract-proposal.js` should create a config object with a consistent order, and `initializeSwingset` should be insensitive to the insertion order of the `config` objects it receives.

cc @dckc for the `extract-proposal.js` side

**Comment by michaelfig** [comment]
> Yeah, `extractCoreProposalBundles()` in

That matches my analysis.  I'll do the work for `extract-proposal.js` to give fully deterministic results.

**Comment by mhofman** [comment]
I'm not following something on the swingset side. Insertion order is a perfectly fine iteration order if the insertion order is stable. I don't think we need to sort every single thing that's being iterated if we can simply propagate the initial order.

In `initializeSwingset`, I do see that `getBundle` is racy, but I can't pinpoint the shared structure that's mutated and would result in non-deterministic insertion. You mention `nameToBundle` being mutated, but I don't see where that happens.

**Comment by michaelfig** [comment]
> I'm not following something on the swingset side. Insertion order is a perfectly fine iteration order if the insertion order is stable. I don't think we need to sort every single thing that's being iterated if we can simply propagate the initial order.

Yeah, I agree with this.  I think the problem is only in cosmic-swingset's `extractCoreProposalBundles`, as that's the first and only time we've auto-generated the SwingSet config structure from a nondeterministic source.  Fixing that should be sufficient.

The fix on the SwingSet side would just be to make it tolerant of nondeterministic config structures, but I'm not convinced that's necessary.

**Comment by michaelfig** [comment]
I've fixed the cosmic-swingset side of this problem (in #5231 above), so I'm closing this issue.

@warner can reopen it if somebody in kernel-land wants to do the busy work of sorting the SwingSet configuration object keys.  That's no longer necessary, and we have other priorities.

**Comment by dckc** [comment]
> @warner can reopen it if somebody in kernel-land wants to ...

I see Warner opened #5229 for that part.
